### PR TITLE
Consider max RF when assessing whether the cluster is over-provisioned

### LINKED ISSUE #2170: Maximum replication factor is not considered when assessing whether the cluster is over-provisioned
While goals such as `RackAwareDistributionGoal` and `RackAwareGoal` take the maximum replication factor into account when assessing over-provisioning, other goals do not. This can lead to provision responses that suggest removing brokers, potentially violating the replication factor of certain topics.

### PR DESCRIPTION
**Pull request by bachmanity1**
This PR resolves #2170.

### CODE REVIEW

**Comment by bachmanity1** [comment]
@CCisGG @efeg @hshi2022 can you please have a look?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/GoalUtils.java [hunk 1]
```diff
@@ -599,4 +600,52 @@ private static double balancePercentageWithMargin(boolean isTriggeredByGoalViola
...
+    if (clusterModel.aliveBrokers().size() < overprovisionedMinBrokers) {
+      return new ProvisionResponse(ProvisionStatus.RIGHT_SIZED);
+    }
+    // when status is OVER_PROVISIONED goal is expected to have exactly 1 recommendation
+    if (provisionResponse.recommendationByRecommender().size() != 1) {
+      throw new IllegalArgumentException(String.format("Expected to have exactly 1 provision recommendation, but got: %d",
+                                                       provisionResponse.recommendationByRecommender().size()));
+    }
+    String goalName = provisionResponse.recommendationByRecommender().keySet().iterator().next();
+    int numBrokersToDrop = provisionResponse.recommendationByRecommender().get(goalName).numBrokers();
```
Comment (CCisGG) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/GoalUtils.java:634]: If it's guaranteed to only have 1 entry in the recommendation map, shall we just use
`provisionResponse.recommendationByRecommender().values().iterator.next().numBrokers()`?

**Review by CCisGG — APPROVED** [review]
Left a non-blocking comment. This change looks reasonble to me. Thanks @bachmanity1 !

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/GoalUtils.java (hunk 1 above)
Comment (bachmanity1) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/GoalUtils.java]: done.

**Comment by bachmanity1** [comment]
@CCisGG can we get this merged?

**Pull request merged**